### PR TITLE
[ENH] Continuing CMI work

### PR DESCRIPTION
**Pull request by adam2392**
Towards #48 

Changes proposed in this pull request:
- implements KSG estimator using scipy KDTree
- implements KSG estimator using an unsupervised tree, which is validated on a Jupyter notebook

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: sktree/experimental/monte_carlo.py
Comment (adam2392) [inline sktree/experimental/monte_carlo.py:1]: @PSSF23 this implements the conditional sampling strategy that can enable both i) bootstrapping and ii) permutation of sample indices conditioned on a conditional array that we want to use to only permute "nearest-neighbors"

File: sktree/experimental/forest.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        else:
+            X = self._validate_data(X, accept_sparse="csc")
+
+        if Z is not None:
+            XZ = np.hstack((X, Z))
+        else:
+            XZ = X
+
+        # if not fitted yet
+        if not hasattr(self.estimator, "estimator_yz_"):
```
Comment (PSSF23) [inline sktree/experimental/forest.py]: Should this line be:
```python
if not hasattr(self, "estimator_yz_"):
```

File: sktree/experimental/monte_carlo.py
Comment (PSSF23) [reply sktree/experimental/monte_carlo.py:1]: LGTM

File: sktree/experimental/forest.py (hunk 1 above)
Comment (adam2392) [reply sktree/experimental/forest.py]: Good catch

**Comment by adam2392** [comment]
I want to add a bunch of unit tests after I ensure that the supervised/unsupervised forests estimate mutual information correctly. 

Rn either there's an issue w/ the simulation, or there's an issue w/ the code.

**Pull request closed without merging**